### ISSUE TITLE
Add generic "acceptance" test for all SeriesSets and ChunkSeriesSets we use.

### ISSUE THREAD

**Issue by bwplotka** [prometheus/prometheus#7151]
With https://github.com/prometheus/prometheus/pull/7059 and https://github.com/prometheus/prometheus/pull/7069 I reduced the number of our series iterators to a very minimum, but still, we have few. Our SeriesSets and ChunkSeriesSet iterator patterns are overall great, but there are some problems: **debuggability and hidden interface requirements**. During development, I spot so many inconsistencies and hidden problems, so it would be nice if we would write a generic helper for our tests to test all new iterators in the same way. This might mean ensuring the same behavior for edge cases and restricting interface behavior, even more, to avoid surprises in the future (: 

I also found some hidden requirements of this iterator that was never mentioned anywhere, e.g:

```
// SeriesSet contains a set of series.
type SeriesSet interface {
	Next() bool
	// [Seen it so I am adding this in #7069] At returns full series. It has to be guaranteed that next or previous
	// iterations do not hold data for the same series.
	At() Series
	Err() error
}

```

This issue is also to agree with what's expected behavior and things we want to test against hugs: (Maybe it's too strict to test if At without Next panics?)

### Things to test against:

* Edge cases:
  * Accessing `At` without `Next` should panic.
  * Accessing `Err` without `Next` should return nil (? I think it would be nice to panic so we will avoid bugs like this) 
  * Accessing `Err` without `Next` should return nil (? I think it would be nice to panic so we will avoid bugs like this) 
  * Accessing `At` after some `Next` more than once.
  * Accessing `At.Iterator()` after some `Next` more than once.
  * Accessing `At.Labels()` after some `Next` more that once.
  * Accessing `At` after `Next` is false should... panic?
  * Invoking `Next` many times after `Next` is false returns false always.
* Normal expand of the iterator +  check if full series were returned!: 

```go
for ss.Next() {
  at := ss.At()
  it := at.Iterator()
  ... at.Labels() // + check that this is unique among ss iterations.
  for it.Next() {
     ...:= it.At()
  }			
  testutil.Ok(t, it.Err())
}		
testutil.Ok(t, ss.Err())
```

* Aggregate first expand of the iterators (PromQL does that, I was so surprised, by code exploded initially :smile: ):

 ```go
var ats []Series
for ss.Next() {
  at = append(at, ss.At())
}		
testutil.Ok(t, ss.Err())

for _, at := range ats {
  it := at.Iterator()
  ... at.Labels()  // + check that this is unique among ss iterations.
  for it.Next() {
     ...:= it.At()
  }			
  testutil.Ok(t, it.Err())
}
```

Does it make sense? @beorn7 @codesome @brian-brazil 

Let's attack this once https://github.com/prometheus/prometheus/pull/7059 and https://github.com/prometheus/prometheus/pull/7069 are merged :hugs:

**Comment by bwplotka** [comment]
Also cc @pstibrany as you had strong opinions on iterators behavior on edge cases :hugs:

**Comment by brian-brazil** [comment]
>It has to be guaranteed that next or previous iterations do not hold data for the same series.

That specific requirement is already stated, as it's a series**set**.

> Accessing At after some Next more than once.

This is completely valid, as are the subsequent ones. You can call At as many times as you like.

> Accessing At after Next is false should... panic?

Panic sounds about right.

> Invoking Next many times after Next is false returns false always.

We could also panic here.

**Comment by bwplotka** [comment]
> That specific requirement is already stated, as it's a seriesset.

I think it's not clear enough, so I assume comment does not harm (:

**Comment by bwplotka** [comment]
>> Accessing At after some Next more than once.

> This is completely valid, as are the subsequent ones. You can call At as many times as you like.

I know but you can easily implement in a way that allocates extra memory or just does not work, thus the need for the test (: 

>> Invoking Next many times after Next is false returns false always.

> We could also panic here.

Can agree.

**Comment by pstibrany** [comment]
>   * Accessing `At` without `Next` should panic.
>   * Accessing `Err` without `Next` should return nil (? I think it would be nice to panic so we will avoid bugs like this)
>   * Accessing `Err` without `Next` should return nil (? I think it would be nice to panic so we will avoid bugs like this)

Interface itself should be explicit that these are *undefined*.

Are you suggesting to change *implementations* inside Prometheus to have this panicking behavior? That sounds like extra work in some cases.

>   * Accessing `At` after some `Next` more than once.
>   * Accessing `At.Iterator()` after some `Next` more than once.
>   * Accessing `At.Labels()` after some `Next` more that once.

These three are valid.

>   * Accessing `At` after `Next` is false should... panic?

Again, undefined. Panic is valid option, but not the only one.

**Comment by pstibrany** [comment]
I don't think it's practical to have all iterators behave in the same way for *undefined behavior*. (Eg. `errSeriesSet` will need lot of extra code to do all this)

**Comment by bwplotka** [comment]
Disagree @pstibrany  **There should not be something like undefined ideally**

Undefined means surprises to me :thinking: Think about cases of bug where a user accidentally use undefined behavior. Suddenly you have weird behavior vs solid panic. 

Fail fast Rule :hugs:

**Comment by bwplotka** [comment]
> Are you suggesting to change implementations inside Prometheus to have this panicking behavior? That sounds like extra work in some cases.

No, those should be fine as you can see I did not mention panic anywhere

**Comment by bwplotka** [comment]
> I don't think it's practical to have all iterators behave in the same way for undefined behavior. (Eg. errSeriesSet will need lot of extra code to do all this)

We should not be afraid of LOCs here - in this case we can reuse single `ErrSeriesSet` created in common place in Prometheus.

**Comment by pstibrany** [comment]
> > > Accessing At after some Next more than once.
> > > Accessing At.Iterator() after some Next more than once.
> > > Accessing At.Labels() after some Next more that once.
> 
> > Are you suggesting to change implementations inside Prometheus to have this panicking behavior? That sounds like extra work in some cases.
> 
> No, those should be fine as you can see I did not mention panic anywhere

You mixed two unrelated quotes here.

Just to be clear about what you're suggesting here: keep the interface as is (eg. you don't want to **require** panicking in those edge cases), but modify implementations **inside** Prometheus according to rules above.

Is that correct? I'm fine with such change, although it seems impractical.

I would object to changes in the interface, where previously unspecified behavior would now require panicking. But that's not what you're proposing, I believe.

**Comment by bwplotka** [comment]
Ok I clarified:

> Clarification for the interface

That's the only change in interface.

> Things to test against

This is what we will require for Prometheus iterators (more strict) and suggest for any implementation, so yea feel free to keep `undefined` behavior if you wish.

**Comment by krajorama** [comment]
Hello from the bug scrub, we still believe that this would be a good way to ensure correctness and consistency. Help wanted.
